### PR TITLE
add 3d content metadata generation

### LINKED ISSUE #6: Produce correct contentMetadata for 3d content type
pending PR noted below 

- [x] depends on sul-dlss/assembly#65 -->  ported to common-accessioning here https://github.com/sul-dlss/common-accessioning/pull/366

- [x] Allow for a new style called "3d" which produces the correct contentMetadata (to be speced). 
see https://consul.stanford.edu/display/chimera/DOR+content+types%2C+resource+types+and+interpretive+metadata

- [x] fix mimetypes issues: sul-dlss/assembly-objectfile#20

### PR DESCRIPTION
**Pull request by peetucket**
addresses #6

also, adjusts default file attributes for PNG as requested by ben

### CODE REVIEW

**Comment by blalbrit** [comment]
A few points for the WIP:

Current contentMD model is correct on: https://argo.stanford.edu/view/druid:bf418ns2590

Devqueue ticket that outlines some thinking about support: https://jirasul.stanford.edu/jira/browse/DEVQUEUE-229?jql=project%20%3D%20DEVQUEUE%20AND%20text%20~%203d

**Comment by peetucket** [comment]
Rebased and updated code to reflect new 3d object content metadata style, which is

```<contentMetadata type="3d">```

and `resource type='3d'` if the resource contains an `'.obj'` file, else `resource type='file'`

File: lib/assembly-objectfile.rb [hunk 1]
```diff
@@ -25,8 +25,9 @@ module Assembly
   FILE_ATTRIBUTES['application/pdf'] = { preserve: 'yes', shelve: 'yes', publish: 'yes' }
   FILE_ATTRIBUTES['plain/text'] = { preserve: 'yes', shelve: 'yes', publish: 'yes' }
   FILE_ATTRIBUTES['text/plain'] = { preserve: 'yes', shelve: 'yes', publish: 'yes' }
-  FILE_ATTRIBUTES['image/png'] = { preserve: 'no', shelve: 'yes', publish: 'yes' }
+  FILE_ATTRIBUTES['image/png'] = { preserve: 'yes', shelve: 'yes', publish: 'no'}
```
Comment (peetucket) [inline lib/assembly-objectfile.rb]: @blalbrit do you know if this change makes sense?  it doesn't seem related to 3d, so not sure why I made it?

Comment (blalbrit) [reply lib/assembly-objectfile.rb]: That change was made, I think, because they didn't want the png files listed in the public content metadata (the files are referenced from the 3d model, but the embed viewer shouldn't try to load them as independent assets.

**Comment by peetucket** [comment]
> Not sure about necessity of changing image/png defaults

confirmed ok change

File: lib/assembly-objectfile.rb (hunk 1 above)
Comment (peetucket) [reply lib/assembly-objectfile.rb]: ok, will leave it in the PR

**Comment by ndushay** [comment]
comparing what is here with this from https://jirasul.stanford.edu/jira/browse/DEVQUEUE-229

- Update https://github.com/sul-dlss/assembly-objectfile/blob/317a421855e14b8a219eb931640a16325561d6ad/lib/assembly-objectfile.rb (and any other location) to manage defaults for obj, ply, threejs, GLTF, etc. type files - PR
- Possibly same codebase... unknown... set mimetype for obj, ply, threejs GLTF, etc. type files that currently fail to provide mimetypes

Does this PR do part or all of the first point above?   Does the second point above need a separate ticket?

File: lib/assembly-objectfile/content_metadata.rb [hunk 2]
```diff
@@ -152,6 +155,14 @@ def self.create_content_metadata(params = {})
                 resource_type_description = resource_has_non_images ? resource_type_descriptions[:object] : resource_type_descriptions[:book]
               when :map
                 resource_type_description = resource_type_descriptions[:map]
+              when :'3d'
+                 resource_extensions = resource_files.collect {|obj| obj.ext}
+                 has_obj_files = resource_extensions.include?('.obj')
```
Comment (ndushay) [inline lib/assembly-objectfile/content_metadata.rb]: per https://jirasul.stanford.edu/jira/browse/DEVQUEUE-229

any obj of type:
- obj
- ply
- threejs
- GLTF
and maybe others, should be covered here, I think.

Comment (peetucket) [reply lib/assembly-objectfile/content_metadata.rb]: added these in latest commit

**Comment by peetucket** [comment]
> comparing what is here with this from [jirasul.stanford.edu/jira/browse/DEVQUEUE-229](https://jirasul.stanford.edu/jira/browse/DEVQUEUE-229)
> 
>     * Update [/lib/assembly-objectfile.rb@`317a421`](https://github.com/sul-dlss/assembly-objectfile/blob/317a421855e14b8a219eb931640a16325561d6ad/lib/assembly-objectfile.rb) (and any other location) to manage defaults for obj, ply, threejs, GLTF, etc. type files - PR
> 
>     * Possibly same codebase... unknown... set mimetype for obj, ply, threejs GLTF, etc. type files that currently fail to provide mimetypes
> 
> 
> Does this PR do part or all of the first point above? Does the second point above need a separate ticket?

This now addresses the first point.  Mimetypes are set from EXIF data or from the mime-types gem (which is updated in this PR).  If we need more intelligence about mimetype setting for unknown or unusual 3D file extensions, that would be a separate ticket, which I will create.

File: lib/assembly-objectfile/content_metadata.rb [hunk 3]
```diff
@@ -152,6 +155,13 @@ def self.create_content_metadata(params = {})
                 resource_type_description = resource_has_non_images ? resource_type_descriptions[:object] : resource_type_descriptions[:book]
               when :map
                 resource_type_description = resource_type_descriptions[:map]
+              when :'3d'
+                 resource_extensions = resource_files.collect {|obj| obj.ext}
+                 if (resource_extensions & VALID_THREE_DIMENSION_EXTENTIONS).empty? # if this resource contains no known 3D file extensions, the resource type is file
```
Comment (ndushay) [inline lib/assembly-objectfile/content_metadata.rb]: why is there no clause like this for any of the other style values?  This doesn't imply, as it does for book, that it is really an "object" instead of a "book"?

File: spec/content_metadata_spec.rb [hunk 4]
```diff
@@ -511,6 +511,27 @@
     expect { described_class.create_content_metadata(druid: TEST_DRUID, objects: objects) }.to raise_error(RuntimeError, "File '#{junk_file}' not found")
   end
 
+  it 'generates valid content metadata for a 3d object with a single obj file and 2 other supporting files' do
```
Comment (ndushay) [inline spec/content_metadata_spec.rb]: doesn't it have 4 files now?

Comment (peetucket) [reply spec/content_metadata_spec.rb]: yes, will adjust name the name of the test

File: lib/assembly-objectfile/content_metadata.rb (hunk 3 above)
Comment (peetucket) [reply lib/assembly-objectfile/content_metadata.rb]: the 3d content metadata spec (https://consul.stanford.edu/display/chimera/DOR+content+types,+resource+types+and+interpretive+metadata) calls for 3d files to be of resource type="3d" and anything else will be of resource type="file"...there are similar types of checks for book (though they go to "object" instead of "file" resource types) but the code looks a little different (it uses the ternary operator instead of an if/then clause)

see https://argo.stanford.edu/view/druid:fx993tk1331 for an existing 3d object like this
